disable leaky_relu_ backward calculation with negative slope

glaringlee opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#33639 disable leaky_relu_ backward calculation with negative slope**

This is to fix the issue #31938
leaky_relu_ (in-place version leakyReLU) can not support backward calculation with negative slope.
The reason is that for in-place (compare to out-of-place) function call, the forward will be saved in autograph node instead of the original input tensor. When doing the backward calculation, we can not figure out what is the sign of the original input based on the forward for leakyReLU, so we can't get a correct backward gradient.
eg. forward is 2, slope is -2. Based on leakyReLU, the original input could be either 2 or -1. 

@albanD The new behavior here is that, if leaky_relu_ backward calculation with a negative slope is detected, we error out and print out an error message that asking the user to call out-of-place version of the forward function instead. 

Removed leaky_relu_backward_out and rrelu_with_noise_backward_out since no one actually use them.

Differential Revision: [D20045735](https://our.internmc.facebook.com/intern/diff/D20045735)

Commit:
disable leaky_relu_ backward calculation with negative slope

[ghstack-poisoned]

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1313,8 +1313,8 @@
-- name: leaky_relu_backward(Tensor grad_output, Tensor self, Scalar negative_slope) -> Tensor
-  grad_output: leaky_relu_backward(grad, self, negative_slope)
+- name: leaky_relu_backward(Tensor grad_output, Tensor self, Scalar negative_slope, bool inplace=False) -> Tensor
+  grad_output: leaky_relu_backward(grad, self, negative_slope, inplace)
```
This leaky relu backward actually corresponds to an out of place leaky relu.
Even if the forward was inplace, the backward never modifies its input inplace, so the backward is an out of place leaky relu.

zou3519 replied:
@glaringlee "inplace" for `leaky_relu_backward` just means "was the forward an in-place function?", right? If so we should rename the argument to `leaky_relu_backward` as `was_forward_inplace` or something

albanD replied:
Yes, we discussed it offline.
The second argument of `leaky_relu_backward` should be renamed from `self` to `self_or_result` and the flag should tell if it is one or the other.

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1346,8 +1346,8 @@
-- name: rrelu_with_noise_backward(Tensor grad_output, Tensor self, Tensor noise, Scalar lower, Scalar upper, bool training) -> Tensor
-  grad_output: rrelu_with_noise_backward(grad, self, noise, lower, upper, training)
+- name: rrelu_with_noise_backward(Tensor grad_output, Tensor self, Tensor noise, Scalar lower, Scalar upper, bool training, bool inplace=False) -> Tensor
+  grad_output: rrelu_with_noise_backward(grad, self, noise, lower, upper, training, inplace)
```
Same as above, this should be false.

albanD reviewed (changes requested):
Looks good, just the double backward logic to check.

gchanan commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -216,15 +216,16 @@ Tensor& rrelu_with_noise_backward_out(
     auto negative = (lower_tensor + upper_tensor) / 2;
     Scalar negative_slope = negative.item();
-    grad_input = at::leaky_relu_backward(grad_output, self, negative_slope);
+    grad_input = at::leaky_relu_backward(grad_output, self, negative_slope, inplace);
```
wait, this shouldn't work you can't just reassign the parameter, you need to use the same memory.

gchanan replied:
do we actually use this out function?

glaringlee replied:
removed

gchanan commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -663,11 +665,25 @@ Tensor & leaky_relu_(
+    bool inplace) {
+  TORCH_CHECK(
+    !inplace || negval.to<double>() > 0.0,
+    "In-place leakyReLu backward calculation is triggered with a negtive slope which is not supported. "
```
negative

zou3519 commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5660,13 +5660,13 @@
     CUDA: legacy::cuda::_thnn_rrelu_with_noise_forward
 
-- func: rrelu_with_noise_backward.grad_input(Tensor grad_output, Tensor self, Tensor noise, Scalar lower, Scalar upper, bool training, *, Tensor(a!) grad_input) -> Tensor(a!)
+- func: rrelu_with_noise_backward.grad_input(Tensor grad_output, Tensor self, Tensor noise, Scalar lower, Scalar upper, bool training, bool inplace=False, *, Tensor(a!) grad_input) -> Tensor(a!)
```
I did a grep through the codebase and I don't see a reason why `rrelu_with_noise_backward.grad_input` or `leaky_relu_backward.grad_input` should exist as operators; they're not used anywhere except for their out-of-place variants. It looks like there were some historical reasons for their existence: back in the day, the THNN CPU/CUDA versions of this function were separate and we used dispatch to call out to them. However, those operators have since been migrated into ATen.

Since we're modifying these lines anyways, we could do the following as a clean-up task:
- delete these two operators from native_functions.yaml
- Call `native::rrelu_with_noise_backward_out` directly from `native::rrelu_with_noise_backward` (and same for `leaky_relu_backward`)

glaringlee replied:
removed out version backward functions

Commit:
Update on "disable leaky_relu_ backward calculation with negative slope"


This is to fix the issue #31938
leaky_relu_ (in-place version leakyReLU) can not support backward calculation with negative slope.
The reason is that for in-place (compare to out-of-place) function call, the forward will be saved in autograph node instead of the original input tensor. When doing the backward calculation, we can not figure out what is the sign of the original input based on the forward for leakyReLU, so we can't get a correct backward gradient.
eg. forward is 2, slope is -2. Based on leakyReLU, the original input could be either 2 or -1. 

@albanD The new behavior here is that, if leaky_relu_ backward calculation with a negative slope is detected, we error out and print out an error message that asking the user to call out-of-place version of the forward function instead. 

Differential Revision: [D20045735](https://our.internmc.facebook.com/intern/diff/D20045735)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -663,22 +644,26 @@ Tensor & leaky_relu_(
-  return grad_input;
-}
-
+// Note: [for leaky_relu_backward and leaky_relu_backward_out]
```
leay_relu_backward_out doesn't exist anymore.

gchanan commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -663,22 +644,26 @@ Tensor & leaky_relu_(
-}
-
+// Note: [for leaky_relu_backward and leaky_relu_backward_out]
+// leakyReLu backward calculation doesn't support in-place call with non-positive slope.
```
maybe say "in-place forward" call to make this clearer.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5599,13 +5599,7 @@
-    CUDA: leaky_relu_backward_out
-
-- func: leaky_relu_backward(Tensor grad_output, Tensor self, Scalar negative_slope) -> Tensor
+- func: leaky_relu_backward(Tensor grad_output, Tensor self, Scalar negative_slope, bool self_is_result=False) -> Tensor
```
why do we have default arguments for backward?  You call backward functions explicitly anyway, i.e. it's not a user API, so we should just always pass every argument.

gchanan commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -5660,13 +5654,7 @@
-    CUDA: rrelu_with_noise_backward_out
-
-- func: rrelu_with_noise_backward(Tensor grad_output, Tensor self, Tensor noise, Scalar lower, Scalar upper, bool training) -> Tensor
+- func: rrelu_with_noise_backward(Tensor grad_output, Tensor self, Tensor noise, Scalar lower, Scalar upper, bool training, bool self_is_result=False) -> Tensor
```
same here

Commit:
Update on "disable leaky_relu_ backward calculation with negative slope"


This is to fix the issue #31938
leaky_relu_ (in-place version leakyReLU) can not support backward calculation with negative slope.
The reason is that for in-place (compare to out-of-place) function call, the forward will be saved in autograph node instead of the original input tensor. When doing the backward calculation, we can not figure out what is the sign of the original input based on the forward for leakyReLU, so we can't get a correct backward gradient.
eg. forward is 2, slope is -2. Based on leakyReLU, the original input could be either 2 or -1. 

@albanD The new behavior here is that, if leaky_relu_ backward calculation with a negative slope is detected, we error out and print out an error message that asking the user to call out-of-place version of the forward function instead. 

Removed leaky_relu_backward_out and rrelu_with_noise_backward_out since no one actually use them.

Differential Revision: [D20045735](https://our.internmc.facebook.com/intern/diff/D20045735)

[ghstack-poisoned]

gchanan commented on aten/src/ATen/native/Activation.cpp:
```diff
@@ -663,22 +644,25 @@ Tensor & leaky_relu_(
+  TORCH_CHECK(
+    !is_result || negval.to<double>() > 0.0,
+    "In-place leakyReLu backward calculation is triggered with a non-positive slope which is not supported. "
+    "This is caused by calling in-place forward function with a non-positive slope, "
```
say something like file an issue (link to github) if you require this support.

Commit:
Update on "disable leaky_relu_ backward calculation with negative slope"


This is to fix the issue #31938
leaky_relu_ (in-place version leakyReLU) can not support backward calculation with negative slope.
The reason is that for in-place (compare to out-of-place) function call, the forward will be saved in autograph node instead of the original input tensor. When doing the backward calculation, we can not figure out what is the sign of the original input based on the forward for leakyReLU, so we can't get a correct backward gradient.
eg. forward is 2, slope is -2. Based on leakyReLU, the original input could be either 2 or -1. 

@albanD The new behavior here is that, if leaky_relu_ backward calculation with a negative slope is detected, we error out and print out an error message that asking the user to call out-of-place version of the forward function instead. 

Removed leaky_relu_backward_out and rrelu_with_noise_backward_out since no one actually use them.

Differential Revision: [D20045735](https://our.internmc.facebook.com/intern/diff/D20045735)

[ghstack-poisoned]

Commit:
Update on "disable leaky_relu_ backward calculation with negative slope"


This is to fix the issue #31938
leaky_relu_ (in-place version leakyReLU) can not support backward calculation with negative slope.
The reason is that for in-place (compare to out-of-place) function call, the forward will be saved in autograph node instead of the original input tensor. When doing the backward calculation, we can not figure out what is the sign of the original input based on the forward for leakyReLU, so we can't get a correct backward gradient.
eg. forward is 2, slope is -2. Based on leakyReLU, the original input could be either 2 or -1. 

@albanD The new behavior here is that, if leaky_relu_ backward calculation with a negative slope is detected, we error out and print out an error message that asking the user to call out-of-place version of the forward function instead. 

Removed leaky_relu_backward_out and rrelu_with_noise_backward_out since no one actually use them.

Differential Revision: [D20045735](https://our.internmc.facebook.com/intern/diff/D20045735)

[ghstack-poisoned]

houseroad commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -53,6 +53,10 @@
     ('_prim::TupleUnpack', datetime.date(2020, 3, 1)),
     ('_aten::format', datetime.date(2020, 3, 1)),
+    ('aten::leaky_relu_backward', datetime.date(2020, 3, 6)),
+    ('aten::leaky_relu_backward.grad_input', datetime.date(2020, 3, 6)),
```
No need to add .grad_input.

houseroad commented on test/backward_compatibility/check_backward_compatibility.py:
```diff
@@ -53,6 +53,10 @@
+    ('aten::leaky_relu_backward', datetime.date(2020, 3, 6)),
+    ('aten::leaky_relu_backward.grad_input', datetime.date(2020, 3, 6)),
+    ('aten::rrelu_with_noise_backward', datetime.date(2020, 3, 6)),
+    ('aten::rrelu_with_noise_backward.grad_input', datetime.date(2020, 3, 6)),
```
ditto

Commit:
Update on "disable leaky_relu_ backward calculation with negative slope"


This is to fix the issue #31938
leaky_relu_ (in-place version leakyReLU) can not support backward calculation with negative slope.
The reason is that for in-place (compare to out-of-place) function call, the forward will be saved in autograph node instead of the original input tensor. When doing the backward calculation, we can not figure out what is the sign of the original input based on the forward for leakyReLU, so we can't get a correct backward gradient.
eg. forward is 2, slope is -2. Based on leakyReLU, the original input could be either 2 or -1. 

@albanD The new behavior here is that, if leaky_relu_ backward calculation with a negative slope is detected, we error out and print out an error message that asking the user to call out-of-place version of the forward function instead. 

Removed leaky_relu_backward_out and rrelu_with_noise_backward_out since no one actually use them.

Differential Revision: [D20045735](https://our.internmc.facebook.com/intern/diff/D20045735)

[ghstack-poisoned]

Commit:
Update on "disable leaky_relu_ backward calculation with negative slope"


This is to fix the issue #31938
leaky_relu_ (in-place version leakyReLU) can not support backward calculation with negative slope.
The reason is that for in-place (compare to out-of-place) function call, the forward will be saved in autograph node instead of the original input tensor. When doing the backward calculation, we can not figure out what is the sign of the original input based on the forward for leakyReLU, so we can't get a correct backward gradient.
eg. forward is 2, slope is -2. Based on leakyReLU, the original input could be either 2 or -1. 

@albanD The new behavior here is that, if leaky_relu_ backward calculation with a negative slope is detected, we error out and print out an error message that asking the user to call out-of-place version of the forward function instead. 

Removed leaky_relu_backward_out and rrelu_with_noise_backward_out since no one actually use them.

Differential Revision: [D20045735](https://our.internmc.facebook.com/intern/diff/D20045735)

[ghstack-poisoned]

houseroad reviewed (approved):
BC part is good to go.

glaringlee commented:
This will break XLA CI test. @ailzhang will update this once XLA side is ready.

ailzhang reviewed (approved):
Thanks for reaching out! I've added a companion PR on xla to fix this issue. 
It's ready to go.
